topology-manager/src/main/java/org/apache/aries/rsa/topologymanager/importer/MultiMap.java

```diff
@@ -19,60 +19,54 @@
 package org.apache.aries.rsa.topologymanager.importer;
 
 import java.util.Collections;
-import java.util.HashMap;
-import java.util.HashSet;
 import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CopyOnWriteArraySet;
 
 /**
- * Minimal implementation of a synchronized map
+ * Minimal implementation of a thread-safe map where each key can have multiple values.
+ *
+ * @param <K> the key type
+ * @param <V> the value type
  */
-public class MultiMap<T> {
+public class MultiMap<K, V> {
 
-    private Map<String, Set<T>> map;
+    private Map<K, Set<V>> map;
     
     public MultiMap() {
-        map = new HashMap<>();
-    }
-    
-    public synchronized void put(String key, T value) {
-        Set<T> values = map.get(key);
-        if (values == null) {
-            values = new HashSet<>();
-            map.put(key, values);
-        }
-        values.add(value);
+        map = new ConcurrentHashMap<>();
     }
     
-    public synchronized Set<T> get(String key) {
-        if (map.containsKey(key)) {
-            return Collections.unmodifiableSet(new HashSet<>(map.get(key)));
-        } else {
-            return Collections.<T>emptySet();
-        }
+    public void put(K key, V value) {
+        map.compute(key, (k, v) -> {
+            if (v == null)
+                v = new CopyOnWriteArraySet<>();
+            v.add(value);
+            return v;
+        });
     }
 
-    public synchronized void remove(String key, T value) {
-        Set<T> values = map.get(key);
-        values.remove(value);
-        if (values.isEmpty()) {
-            map.remove(key);
-        }
+    // the returned set should not be modified directly
+    public Set<V> get(K key) {
+        return map.getOrDefault(key, Collections.emptySet());
     }
 
-    public synchronized Set<String> keySet() {
-        return Collections.unmodifiableSet(new HashSet<>(map.keySet()));
+    public void remove(K key, V value) {
+        map.compute(key, (k, v) -> v != null && v.remove(value) && v.isEmpty() ? null : v);
     }
 
-    public synchronized void remove(T toRemove) {
-        // Use copy of keySet, as subsequent remove may modify the keySet itself
-        Set<String> keys = new HashSet<>(map.keySet());
-        for (String key : keys) {
-            remove(key, toRemove);
+    public void remove(V value) {
+        for (K key : map.keySet()) {
+            remove(key, value);
         }
     }
 
-    public synchronized void clear() {
+    public Set<K> keySet() {
+        return map.keySet();
```

Comment:
As values should only be added/removed through this MultiMap implementation, I would propose to wrap the result of getOrDefault with Collection.unmodifiableSet()

Reply:
These are just the keys of the whole map, not the values or inner sets - no problem if someone wants to remove a key. Add is not supported in this view.

Reply:
Removing an element of the keySet will also remove the associated value of the underlying map. There may be use-cases where you'd want that, but I would add these as functions of the MultiMap and not have using code modify the MultiMap through any returned collection.

Reply:
That is precisely the contract of the general Map.keySet() interface, so I think it would be the expected and intuitive behavior. The inner sets need protection from external modification to ensure thread-safety (when they become empty and are removed from the map), but the map itself can be delegated to elsewhere, as it behaves correctly, is thread-safe, has less code and less garbage... but we can let @cschneider decide either way :-)